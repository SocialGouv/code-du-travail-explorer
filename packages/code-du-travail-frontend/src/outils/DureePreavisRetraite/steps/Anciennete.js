import { Input } from "@socialgouv/cdtn-ui";
import { format, isAfter } from "date-fns";
import PropTypes from "prop-types";
import Engine from "publicodes";
import React from "react";
import { Field } from "react-final-form";

import { parse } from "../../common/date";
import { Error, ErrorComputedField } from "../../common/ErrorField";
import { SectionTitle } from "../../common/stepStyles";
import { TextQuestion } from "../../common/TextQuestion";
import { isDate, isNumber } from "../../common/validators";
import rules from "../modele/index";

function validate({ dateEntree, dateSortie, dateNotification }) {
  const errors = {};
  const dEntree = parse(dateEntree);
  const dSortie = parse(dateSortie);
  const dNotification = parse(dateNotification);

  if (dateEntree && dateSortie && isAfter(dEntree, dSortie)) {
    errors.dateSortie = (
      <>
        La date de sortie doit se situer après le
        <strong> {format(dEntree, "dd MMMM yyyy")}</strong>
      </>
    );
  }

  if (dateNotification && dateSortie && isAfter(dNotification, dSortie)) {
    errors.dateNotification = `La date de notification doit se situer avant la date de sortie`;
  }
  if (dateNotification && dateEntree && isAfter(dEntree, dNotification)) {
    errors.dateNotification = `La date de notification doit se situer après la date d’entrée`;
  }
  return errors;
}

function StepAnciennete({ form }) {
  const engine = new Engine(rules);
  engine.setSituation({
    "contrat salarié . convention collective": `'IDCC${
      form.getState().values.ccn.num
    }'`,
  });
  const result = engine.evaluate("contrat salarié . ancienneté");

  const absences = Object.keys(result.missingVariables)
    .filter((item) => {
      return item.includes(" . absences . ");
    })
    .map((item) => {
      const rule = engine.getRule(item);
      return {
        title: rule.title,
        description: rule.rawNode.description,
        id: item,
      };
    });
  return (
    <>
      <SectionTitle>Dates d’entrée et de sortie de l’entreprise</SectionTitle>
      <TextQuestion
        name="contrat salarié . date embauche"
        label="Quelle est la date d’entrée dans l’entreprise&nbsp;?"
        inputType="date"
        validate={isDate}
        placeholder=" jj/mm/yyyy" // placeholder for safari desktop which does not support input type date
      />
      <TextQuestion
        name="contrat salarié . date rupture"
        label="Quelle est la date de sortie de l’entreprise&nbsp;?"
        inputType="date"
        validate={isDate}
        validateOnChange
        placeholder=" jj/mm/yyyy" // placeholder for safari desktop which does not support input type date
      />
      <ErrorComputedField name="anciennete" />
      <SectionTitle>Période d’absence prolongée</SectionTitle>
      <div
        style={{
          display: "flex",
          flexDirection: "column",
        }}
      >
        {absences.map((item) => (
          <div
            key={item}
            style={{
              alignItems: "center",
              display: "flex",
              justifyContent: "space-between",
              margin: 10,
            }}
          >
            <span
              style={{
                flex: "2",
              }}
              title={item.description}
            >
              {item.title}
            </span>
            <Field
              style={{ flex: "1" }}
              name={`${item.id}`}
              validate={isNumber}
              subscription={{
                error: true,
                invalid: true,
                touched: true,
                value: true,
              }}
              render={({ input, meta: { touched, error, invalid } }) => (
                <>
                  <Input
                    {...input}
                    id={`${item.id}`}
                    type="number"
                    invalid={touched && invalid}
                  />
                  {error && touched && invalid && <Error>{error}</Error>}
                </>
              )}
            />
          </div>
        ))}
      </div>
    </>
  );
}

StepAnciennete.validate = validate;

StepAnciennete.propTypes = {
  form: PropTypes.object.isRequired,
};

export { StepAnciennete };
