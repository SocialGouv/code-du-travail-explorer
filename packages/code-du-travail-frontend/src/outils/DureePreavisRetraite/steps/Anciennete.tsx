import React from "react";

import { TextQuestion } from "../../common/TextQuestion";
import { isPositiveNumber } from "../../common/validators";

function AncienneteStep() {
  return (
    <>
      <TextQuestion
        name="contrat salarié - ancienneté"
        label="Quelle est votre ancienneté dans l’entreprise en mois&nbsp;?"
        inputType="number"
        validate={isPositiveNumber}
        placeholder="0"
      />
    </>
  );
}

export { AncienneteStep };
