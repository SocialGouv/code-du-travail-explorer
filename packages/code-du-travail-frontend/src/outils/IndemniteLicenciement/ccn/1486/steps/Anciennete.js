import React from "react";
import { CATEGORIE_KEY } from "./Categorie";
import { AncienneteETAMIC } from "./AncienneteETAMIC";
import { AncienneteCE } from "./AncienneteCE";

export const Anciennete = ({ form }) => {
  const categorie = form.getState().values[CATEGORIE_KEY];
  if (categorie === "CEI" || categorie === "CENI") {
    return <AncienneteCE form={form} />;
  }
  return <AncienneteETAMIC form={form} />;
};
