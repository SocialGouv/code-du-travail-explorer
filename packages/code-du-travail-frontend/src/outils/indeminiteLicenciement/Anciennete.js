import React from "react";
import PropTypes from "prop-types";
import { Section } from "@cdt/ui";

import { inputStyle } from "./index";
import { PrevNextStepper } from "./PrevNextStepper";

class Anciennete extends React.Component {
  static propTypes = {
    value: PropTypes.number,
    onChange: PropTypes.func.isRequired,
    onComplete: PropTypes.func.isRequired,
    nextDisabled: PropTypes.bool
  };

  static defaultProps = {
    value: 0,
    nextDisabled: false
  };

  render() {
    const { value, onChange, onComplete, nextDisabled } = this.props;
    return (
      <Section light>
        <React.Fragment>
          <h2>Quelle est votre anciennet√© en mois ?</h2>
          <div style={{ fontSize: "2em" }}>
            <input
              type="number"
              onFocus={e => {
                if (e.target.value === "0") {
                  e.target.value = "";
                }
              }}
              onChange={e => onChange(parseFloat(e.target.value) || 0)}
              value={value}
              style={{ width: 100, ...inputStyle }}
            />{" "}
            mois
            <PrevNextStepper onNext={onComplete} nextDisabled={nextDisabled} />
          </div>
        </React.Fragment>
      </Section>
    );
  }
}

export { Anciennete };
