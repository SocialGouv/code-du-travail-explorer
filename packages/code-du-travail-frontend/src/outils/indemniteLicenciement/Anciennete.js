import React from "react";
import PropTypes from "prop-types";
import { Container, Section } from "@cdt/ui";

import { inputStyle } from "./stepStyles";
import { PrevNextStepper } from "./PrevNextStepper";

class Anciennete extends React.Component {
  static propTypes = {
    value: PropTypes.number,
    onChange: PropTypes.func.isRequired,
    onPrevious: PropTypes.func.isRequired,
    onNext: PropTypes.func.isRequired,
    nextDisabled: PropTypes.bool
  };

  static defaultProps = {
    value: 0,
    nextDisabled: false
  };

  render() {
    const { value, onChange, onNext, onPrevious, nextDisabled } = this.props;
    return (
      <React.Fragment>
        <Section light>
          <React.Fragment>
            <h2>Quelle est votre anciennet√© en mois ?</h2>
            <div style={{ fontSize: "2em" }}>
              <input
                type="number"
                onFocus={e => {
                  if (e.target.value === "0") {
                    e.target.value = "";
                  }
                }}
                onChange={e => onChange(parseFloat(e.target.value) || 0)}
                value={value}
                style={{ width: 180, ...inputStyle }}
              />{" "}
              mois
            </div>
          </React.Fragment>
        </Section>
        <Container>
          <PrevNextStepper
            onPrev={onPrevious}
            onNext={onNext}
            nextDisabled={nextDisabled}
          />
        </Container>
      </React.Fragment>
    );
  }
}

export { Anciennete };
