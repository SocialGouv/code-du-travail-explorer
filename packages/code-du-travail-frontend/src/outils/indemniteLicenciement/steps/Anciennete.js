import React from "react";
import { Container } from "@cdt/ui";
import { Field } from "react-final-form";
import { OnChange } from "react-final-form-listeners";
import createDecorator from "final-form-calculate";
import { isAfter, differenceInMonths, format } from "date-fns";

import { YesNoQuestion } from "../components/YesNoQuestion";
import { TextQuestion } from "../components/TextQuestion";
import { AbsencePeriods, motifs } from "../components/AbsencePeriods";

function validate({
  dateEntree,
  dateSortie,
  dateNotification,
  absencePeriods = []
}) {
  const errors = {};
  const dEntree = new Date(dateEntree);
  const dSortie = new Date(dateSortie);
  const dNotification = new Date(dateNotification);

  if (dateEntree && dateSortie && isAfter(dEntree, dSortie)) {
    errors.dateSortie = (
      <>
        La date de sortie doit se situer apres le
        <strong> {format(dEntree, "DD MMMM YYYY")}</strong>
      </>
    );
  }

  const totalAbsence =
    (absencePeriods || [])
      .filter(period => Boolean(period.duration))
      .reduce((total, item) => {
        const motif = motifs.find(motif => motif.label === item.type);
        return total + item.duration * motif.value;
      }, 0) / 12;

  if (
    dateEntree &&
    dateSortie &&
    isAfter(new Date("2017-09-27"), dateSortie) &&
    differenceInMonths(dNotification, dEntree) - totalAbsence < 12
  ) {
    errors.dateSortie =
      "L’indemnité de licenciement est dûe au-delà de 12 mois d'ancienneté";
  }
  if (
    dateEntree &&
    dateSortie &&
    isAfter(dateSortie, new Date("2017-09-27")) &&
    differenceInMonths(dNotification, dEntree) - totalAbsence < 8
  ) {
    errors.dateSortie =
      "L’indemnité de licenciement est dûe au-delà de 8 mois d'ancienneté";
  }

  if (dateNotification && dateSortie && isAfter(dNotification, dSortie)) {
    errors.dateNotification = `La date de notification doit se situer avant la date de sortie`;
  }
  return errors;
}

function StepAnciennete({ form }) {
  return (
    <Container narrow nopadding>
      <TextQuestion
        name="dateEntree"
        label="Quelle est votre date d'entrée dans l'entreprise&nbsp;?"
        inputType="date"
      />
      <TextQuestion
        name="dateNotification"
        label="À quelle date votre licenciement vous a-t-il été notifié&nbsp;?"
        inputType="date"
      />
      <TextQuestion
        name="dateSortie"
        label="Quelle est votre date de sortie de l'entreprise (incluant la durée de votre préavis)&nbsp;?"
        inputType="date"
      />
      <br />
      <br />
      <YesNoQuestion
        name="hasAbsenceProlonge"
        label="Avez-vous eu des périodes d'absence de plus d'un mois au cours
        de votre contrat&nbsp;?"
      />
      <OnChange name="hasAbsenceProlonge">
        {hasAbsenceProlonge => {
          hasAbsenceProlonge
            ? form.change("absencePeriods", [
                {
                  type: "Absence pour maladie non professionnelle",
                  duration: null
                }
              ])
            : form.change("absencePeriods", []);
        }}
      </OnChange>
      <br />
      <br />
      <Field name="hasAbsenceProlonge">
        {({ input }) =>
          input.value === true ? <AbsencePeriods name="absencePeriods" /> : null
        }
      </Field>
      <OnChange name="absencePeriods">
        {absencePeriods => {
          if (absencePeriods.length === 0) {
            form.change("hasAbsenceProlonge", false);
          }
        }}
      </OnChange>
    </Container>
  );
}

StepAnciennete.validate = validate;
/**
 * The decorator here is used to compute the ancienneté value
 * based on the data provided by the user
 * decorator can only be used in initialSteps since final-form do not allows
 * decorator to be added once the form is created
 */
StepAnciennete.decorator = createDecorator({
  field: /date|absencePeriods/,
  updates: {
    anciennete: (_, { dateEntree, dateSortie, absencePeriods = [] }) => {
      const dEntree = new Date(dateEntree);
      const dSortie = new Date(dateSortie);
      // on calcule totalAbsence en mois par année (ex: 12mois = 1)
      // pour pouvoir ensuite le retranché de l'anciennété qui est aussi en mois par année
      const totalAbsence =
        (absencePeriods || [])
          .filter(period => Boolean(period.duration))
          .reduce((total, item) => {
            const motif = motifs.find(motif => motif.label === item.type);
            return total + item.duration * motif.value;
          }, 0) / 12;
      return (differenceInMonths(dSortie, dEntree) - totalAbsence) / 12;
    }
  }
});

export { StepAnciennete };
