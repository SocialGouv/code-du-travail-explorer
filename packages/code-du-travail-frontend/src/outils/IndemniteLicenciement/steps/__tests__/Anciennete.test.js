import React from "react";
import { render, fireEvent } from "@testing-library/react";
import arrayMutators from "final-form-arrays";
import { Form } from "react-final-form";
import { StepAnciennete, computeAnciennete } from "../Anciennete";

function renderForm(data) {
  return render(
    <Form
      mutators={{
        ...arrayMutators
      }}
      validate={StepAnciennete.validate}
      initialValues={{ ...data }}
      onSubmit={jest.fn()}
    >
      {({ form }) => <StepAnciennete form={form} />}
    </Form>
  );
}

describe("<Anciennete />", () => {
  it("should render", () => {
    const { container } = renderForm({});
    expect(container).toMatchSnapshot();
  });

  it("should display error if dateEntre after dateSortie", () => {
    const { getByLabelText, getByText } = renderForm({
      dateEntree: "2018-04-02"
    });
    const dateSortie = getByLabelText(/date de sortie/i);
    fireEvent.change(dateSortie, { target: { value: "2018-03-01" } });
    expect(
      getByText(/la date de sortie doit se situer après/i)
    ).toMatchSnapshot();
  });

  it("should display error if dateEntree after dateNotif", () => {
    const { container, getByLabelText } = renderForm({
      dateEntrer: "2018-05-01"
    });
    const dateNotif = getByLabelText(/licenciement/i);
    fireEvent.change(dateNotif, { target: { value: "2018-04-02" } });
    expect(container.querySelector(".alert")).toMatchSnapshot();
  });

  it("should display error if dateNotif after dateSortie", () => {
    const { container, getByLabelText } = renderForm({
      dateSortie: "2018-04-01"
    });
    const dateNotif = getByLabelText(/licenciement/i);
    fireEvent.change(dateNotif, { target: { value: "2018-05-02" } });
    expect(container.querySelector(".alert")).toMatchSnapshot();
  });

  it("should display error if anciennté < 8mois", () => {
    const { getByLabelText, getByText } = renderForm({
      dateEntree: "2018-04-02",
      dateNotification: "2018-09-02"
    });
    const dateSortie = getByLabelText(/date de sortie/i);
    fireEvent.change(dateSortie, { target: { value: "2018-12-31" } });
    expect(getByText(/est dûe au-delà de 8 mois/i)).toMatchSnapshot();
  });
});

describe("computeAncienneté", () => {
  it("should compute ancienneté", () => {
    expect(
      computeAnciennete({ dateEntree: "2017-04-01", dateSortie: "2018-04-01" })
    ).toEqual(1);
  });
  it("should compute ancienneté with periods of absence", () => {
    expect(
      computeAnciennete({
        dateEntree: "2017-04-01",
        dateSortie: "2018-04-01",
        absencePeriods: [{ type: "Congés sans solde", duration: "6" }]
      })
    ).toEqual(0.5);
  });
  it("should compute ancienneté with periods of absence divided by two for Congé parental", () => {
    expect(
      computeAnciennete({
        dateEntree: "2016-04-01",
        dateSortie: "2018-04-01",
        absencePeriods: [{ type: "Congé parental d'éducation", duration: "6" }]
      })
    ).toEqual(1.75);
  });
});
