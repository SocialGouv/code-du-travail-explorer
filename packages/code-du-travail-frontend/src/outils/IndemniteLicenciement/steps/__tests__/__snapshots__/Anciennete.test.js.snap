// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anciennete /> should render 1`] = `
.c4 {
  width: auto;
  padding: 1rem 0.4rem;
  font-size: 1rem;
  line-height: inherit;
  text-align: left;
  border: 1px solid #bbcadf;
  border-color: #bbcadf;
  border-radius: 0.6rem;
}

.c4::-webkit-outer-spin-button,
.c4::-webkit-inner-spin-button {
  margin: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 2em;
  padding: 0;
  font-size: 1.8rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-bottom: 2rem;
}

.c0 {
  margin: 0 0 1rem 0;
  color: #2f3b6c;
  font-weight: normal;
  font-size: 2.8rem;
  font-family: "Merriweather",serif;
  line-height: 1.25;
  margin-top: 3.2rem;
  margin-bottom: 2rem;
}

.c1 {
  display: block;
  margin-top: 2rem;
  margin-bottom: 1rem;
  font-size: 1.8rem;
  cursor: pointer;
}

.c5 {
  display: block;
  margin-top: 2rem;
  margin-bottom: 1rem;
  font-size: 1.8rem;
  cursor: default;
}

.c2 {
  display: inline-block;
  margin-left: 1rem;
  color: #eb5757;
  font-weight: 700;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 2rem;
}

<div>
  <form>
    <h2
      class="c0"
    >
      Dates d’entrée et de sortie de l’entreprise
    </h2>
    <label
      class="c1"
      for="1"
    >
      Quelle est la date d’entrée dans l’entreprise ?
      <span
        aria-label="champs obligatoire"
        class="c2"
      >
        *
      </span>
    </label>
    <div
      class="c3"
    >
      <input
        class="c4"
        id="1"
        name="dateEntree"
        placeholder=" jj/mm/yyyy"
        type="date"
        value=""
      />
    </div>
    <label
      class="c1"
      for="2"
    >
      Quelle est la date de notification du licenciement ?
      <span
        aria-label="champs obligatoire"
        class="c2"
      >
        *
      </span>
    </label>
    <div
      class="c3"
    >
      <input
        class="c4"
        id="2"
        name="dateNotification"
        placeholder=" jj/mm/yyyy"
        type="date"
        value=""
      />
    </div>
    <label
      class="c1"
      for="3"
    >
      Quelle est la date de sortie de l’entreprise (incluant la durée du préavis) ?
      <span
        aria-label="champs obligatoire"
        class="c2"
      >
        *
      </span>
    </label>
    <div
      class="c3"
    >
      <input
        class="c4"
        id="3"
        name="dateSortie"
        placeholder=" jj/mm/yyyy"
        type="date"
        value=""
      />
    </div>
    <h2
      class="c0"
    >
      Période d’absence prolongée (plus d’un mois)
    </h2>
    <p
      class="c5"
    >
      Y a-t-il eu des absences de plus d’un mois durant le contrat de travail ?
      <span
        aria-label="champs obligatoire"
        class="c2"
      >
        *
      </span>
    </p>
    <div
      class="c6"
    >
      <label
        class="c7"
      >
        <input
          name="hasAbsenceProlonge"
          type="radio"
          value="true"
        />
        <span>
          Oui
        </span>
      </label>
      <label
        class="c7"
      >
        <input
          name="hasAbsenceProlonge"
          type="radio"
          value="false"
        />
        <span>
          Non
        </span>
      </label>
    </div>
    
    
    <button
      data-testid="nextBt"
    >
      suivant
    </button>
  </form>
</div>
`;
